plugin/src/Internal/Compile.hs

```diff
@@ -1,11 +1,94 @@
+{-# language ViewPatterns, CPP #-}
+
 module Internal.Compile where
 
-import Control.Monad.IO.Class (liftIO)
-import GHC (Ghc, Phase, getSession)
-import GHC.Driver.Phases (StopPhase (NoStop))
-import GHC.Driver.Pipeline (oneShot)
+import Data.Maybe (isJust)
+import GHC (DynFlags (..), Ghc, GhcException (..), GhcMonad (..), HscEnv, Phase)
+import GHC.Driver.Backend (backendGeneratesCode)
+import GHC.Driver.Env (HscEnv (..), hscUpdateFlags)
+import GHC.Driver.Phases (Phase (..), StopPhase (..), startPhase)
+import GHC.Driver.Pipeline (
+  TPhase,
+  TPipelineClass,
+  asPipeline,
+  cmmCppPipeline,
+  cmmPipeline,
+  fullPipeline,
+  llvmLlcPipeline,
+  llvmManglePipeline,
+  llvmPipeline,
+  mkPipeEnv,
+  runPipeline,
+  viaCPipeline,
+  )
+import GHC.Driver.Pipeline.Monad (PipeEnv (..), PipelineOutput (..))
+import GHC.Driver.Session (augmentByWorkingDirectory, isNoLink, outputFile)
+import GHC.Runtime.Loader (initializePlugins)
+import GHC.Types.SourceFile (HscSource)
+import GHC.Unit.Home.ModInfo (HomeModLinkable (..))
+import GHC.Utils.Monad (MonadIO (..), unlessM)
+import GHC.Utils.Panic (panic, throwGhcExceptionIO)
+import Internal.Cache (ModuleArtifacts (..), Target (..))
+import System.Directory (doesFileExist)
+
+type P m = TPipelineClass TPhase m
+
+pipelineOneshot :: P m => PipeEnv -> HscEnv -> FilePath -> m (Maybe ModuleArtifacts)
```

Comment:
in this module, what are reusable parts from the current GHC API?
It is good to have some note and upstreaming plan for that.